### ISSUE TITLE
Adding audio payload with unsupported clock-rate breaks pipeline

### ISSUE THREAD

**Issue by ikonst** [EricssonResearch/openwebrtc#141]
Adding an audio payload with unsupported clock-rate breaks the pipeline. For instance, GstOpusEnc supports rates { 48000, 24000, 16000, 12000, 8000 }.

```c
owr_audio_payload_new(
   OWR_CODEC_TYPE_OPUS,
  100, // dynamic RTP payload type
   44100, // clock rate
   1 // channels
);
```

This will create a broken pipeline, whose source-bin cannot link with the transport-agent. This becomes known late, and with a nondescript error message.

While I think implementing our own parameter checking for payloads, in addition to the Gst caps negotiation, would be wrong since it's duplication, but I think the errors could be friendlier.

**Comment by ikonst** [comment]
Here are Graphviz files of:
* a working pipeline (good-opus48000.dot) with clock-rate=48000
* a broken pipeline (bad-opus44100.dot) with clock-rate=44100

https://drive.google.com/folderview?id=0B359dq-Bj8zdfkdma2tEVEVIcFZVLUxldDNKQWp1bWZ2S0NLNHgyWDdoeGcwS1drQTVOZEU&usp=sharing

**Comment by superdump** [comment]
We need to implement some sensible error propagation interface so that useful information can easily be had in case of such errors. In this case we can at least document the values expected to work.

**Comment by sdroege** [comment]
Yeah, currently there is no sensible error reporting at all and internally most errors are just ignored or lead to an assertion. That needs someone to sit down and go over all the code base for possible errors and then come up with an API.

**Comment by ikonst** [comment]
In the case of Opus, actually, it appears that GstRtpOPUSPay is even more restrictive, supporting only clock-rate of 48000.

In the bad pipeline, I can see the following errors:
* send-rtp-capsfilter requires clock-rate=44100 and therefore pay_rtpopuspay cannot connect to it
* encoder_opusenc supports rates { 48000, 24000, 16000, 12000, 8000 } and therefore cannot connect to the source-output-queue (which is filtered to 44100)

I understand that we only see these problems at link time?

**Comment by sdroege** [comment]
Is this still a problem? This should be automatically solveable, and with latest GIT of GStreamer also seems to be. Compare:

gst-launch-1.0 audiotestsrc ! "audio/x-raw,rate=44100" ! audioresample ! opusenc ! rtpopuspay ! fakesink
gst-launch-1.0 audiotestsrc ! "audio/x-raw,rate=44100" ! opusenc ! rtpopuspay ! fakesink

The second is failing, the first is properly resampling to 48kHz before passing data to opusenc. And there should be an audioresample before opusenc in Owr.

**Comment by superdump** [comment]
@ikonst is this still a problem?

**Comment by ikonst** [comment]
I'll test it to see if it's still happening, thanks for the reminder.
